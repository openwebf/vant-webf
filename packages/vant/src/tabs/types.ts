import type { ComponentPublicInstance, ComputedRef } from 'vue';
import type { Numeric } from '../utils';
import type { TabsProps } from './Tabs';

export type TabsType = 'line' | 'card';

export type TabsClickTabEventParams = {
  name: Numeric;
  title: string;
  event: MouseEvent;
  disabled: boolean;
};

export type TabsProvide = {
  id: string;
  props: TabsProps;
  setLine: () => void;
  onRendered: (name: Numeric, title?: string) => void;
  scrollIntoView: (immediate?: boolean) => void;
  currentName: ComputedRef<Numeric | undefined>;
};

export type TabsExpose = {
  resize: () => void;
  scrollTo: (name: Numeric) => void;
};

export type TabsInstance = ComponentPublicInstance<TabsProps, TabsExpose>;
