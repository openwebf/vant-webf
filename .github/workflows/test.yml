name: CI

on:
  push:
    branches:
      - '**'

  pull_request:
    branches:
      - dev

  workflow_dispatch:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run linter
      run: npm run lint

  test:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Run test cases
      run: npm test

    - name: Upload coverage to Codecov  
      uses: codecov/codecov-action@v1
      with:
        token: ${{ secrets.CODECOV_TOKEN }}

  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2

    - name: Install pnpm
      run: npm i pnpm@7 -g

    - uses: actions/setup-node@v2
      with:
        node-version: '16'
        cache: 'pnpm'

    - name: Install dependencies
      run: pnpm install

    - name: Build
      run: npm run build
