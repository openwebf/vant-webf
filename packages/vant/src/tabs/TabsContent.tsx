import { ref, watch, onMounted, defineComponent } from 'vue';
import { numericProp, makeRequiredProp, createNamespace } from '../utils';
import { Swipe, SwipeInstance } from '../swipe';
import { useExpose } from '../composables/use-expose';

const [name, bem] = createNamespace('tabs');

export default defineComponent({
  name,

  props: {
    count: makeRequiredProp(Number),
    inited: Boolean,
    animated: Boolean,
    duration: makeRequiredProp(numericProp),
    swipeable: Boolean,
    lazyRender: Boolean,
    currentIndex: makeRequiredProp(Number),
  },

  emits: ['change'],

  setup(props, { emit, slots }) {
    const swipeRef = ref<SwipeInstance>();

    const onChange = (index: number) => emit('change', index);

    const renderChildren = () => {
      const Content = slots.default?.();

      if (props.animated || props.swipeable) {
        return (
          <Swipe
            ref={swipeRef}
            loop={false}
            class={bem('track')}
            duration={+props.duration * 1000}
            touchable={props.swipeable}
            lazyRender={props.lazyRender}
            showIndicators={false}
            onChange={onChange}
          >
            {Content}
          </Swipe>
        );
      }

      return Content;
    };

    const swipeToCurrentTab = (index: number) => {
      const swipe = swipeRef.value;
      if (swipe && swipe.state.active !== index) {
        swipe.swipeTo(index, { immediate: !props.inited });
      }
    };

    watch(() => props.currentIndex, swipeToCurrentTab);

    onMounted(() => {
      swipeToCurrentTab(props.currentIndex);
    });

    useExpose({ swipeRef });

    return () => (
      <div
        class={bem('content', {
          animated: props.animated || props.swipeable,
        })}
      >
        {renderChildren()}
      </div>
    );
  },
});
