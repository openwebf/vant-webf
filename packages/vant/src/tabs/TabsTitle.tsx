import { computed, defineComponent, type CSSProperties } from 'vue';
import { isDef, truthProp, numericProp, createNamespace } from '../utils';
import { Badge } from '../badge';

const [name, bem] = createNamespace('tab');

export default defineComponent({
  name,

  props: {
    id: String,
    dot: Boolean,
    type: String,
    color: String,
    title: String,
    badge: numericProp,
    shrink: Boolean,
    isActive: Boolean,
    disabled: Boolean,
    controls: String,
    scrollable: Boolean,
    activeColor: String,
    inactiveColor: String,
    showZeroBadge: truthProp,
  },

  setup(props, { slots }) {
    const style = computed(() => {
      const style: CSSProperties = {};
      const { type, color, disabled, isActive, activeColor, inactiveColor } =
        props;

      const isCard = type === 'card';

      // card theme color
      if (color && isCard) {
        style.borderColor = color;

        if (!disabled) {
          if (isActive) {
            style.backgroundColor = color;
          } else {
            style.color = color;
          }
        }
      }

      const titleColor = isActive ? activeColor : inactiveColor;
      if (titleColor) {
        style.color = titleColor;
      }

      return style;
    });

    const renderText = () => {
      const Text = (
        <span class={bem('text', { ellipsis: !props.scrollable })}>
          {slots.title ? slots.title() : props.title}
        </span>
      );

      if (props.dot || (isDef(props.badge) && props.badge !== '')) {
        return (
          <Badge
            dot={props.dot}
            content={props.badge}
            showZero={props.showZeroBadge}
          >
            {Text}
          </Badge>
        );
      }

      return Text;
    };

    return () => (
      <div
        id={props.id}
        role="tab"
        class={[
          bem([
            props.type,
            {
              grow: props.scrollable && !props.shrink,
              shrink: props.shrink,
              active: props.isActive,
              disabled: props.disabled,
            },
          ]),
        ]}
        style={style.value}
        tabindex={props.disabled ? undefined : props.isActive ? 0 : -1}
        aria-selected={props.isActive}
        aria-disabled={props.disabled || undefined}
        aria-controls={props.controls}
      >
        {renderText()}
      </div>
    );
  },
});
