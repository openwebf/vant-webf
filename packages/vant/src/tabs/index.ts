import { withInstall } from '../utils';
import _Tabs, { TabsProps } from './Tabs';

export const Tabs = withInstall(_Tabs);
export default Tabs;
export type { TabsProps };
export type { TabsType, TabsInstance } from './types';

declare module 'vue' {
  export interface GlobalComponents {
    VanTabs: typeof Tabs;
  }
}
