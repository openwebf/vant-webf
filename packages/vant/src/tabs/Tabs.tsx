import {
  ref,
  watch,
  computed,
  reactive,
  nextTick,
  onActivated,
  defineComponent,
  getCurrentInstance,
  type PropType,
  type InjectionKey,
  type CSSProperties,
  type ExtractPropTypes,
  type ComponentPublicInstance,
} from 'vue';

// Utils
import {
  pick,
  isDef,
  addUnit,
  isHidden,
  unitToPx,
  truthProp,
  numericProp,
  windowWidth,
  getElementTop,
  makeStringProp,
  callInterceptor,
  createNamespace,
  makeNumericProp,
  setRootScrollTop,
  BORDER_TOP_BOTTOM,
  type Numeric,
  type Interceptor,
  type ComponentInstance,
} from '../utils';
import { scrollLeftTo, scrollTopTo } from './utils';

// Composables
import {
  useRect,
  useChildren,
  useScrollParent,
  useEventListener,
  onMountedOrActivated,
} from '@vant/use';
import { useId } from '../composables/use-id';
import { route, RouteProps } from '../composables/use-route';
import { useRefs } from '../composables/use-refs';
import { useExpose } from '../composables/use-expose';
import { onPopupReopen } from '../composables/on-popup-reopen';

// Components
import { Sticky } from '../sticky';
import TabsTitle from './TabsTitle';
import TabsContent from './TabsContent';

// Types
import type { TabsProvide, TabsType } from './types';

const [name, bem] = createNamespace('tabs');

const tabsProps = {
  type: makeStringProp<TabsType>('line'),
  color: String,
  border: Boolean,
  sticky: Boolean,
  shrink: Boolean,
  active: makeNumericProp(0),
  duration: makeNumericProp(0.3),
  animated: Boolean,
  ellipsis: truthProp,
  swipeable: Boolean,
  scrollspy: Boolean,
  offsetTop: makeNumericProp(0),
  background: String,
  lazyRender: truthProp,
  lineWidth: numericProp,
  lineHeight: numericProp,
  beforeChange: Function as PropType<Interceptor>,
  swipeThreshold: makeNumericProp(5),
  titleActiveColor: String,
  titleInactiveColor: String,
};

export type TabsProps = ExtractPropTypes<typeof tabsProps>;

export const TABS_KEY: InjectionKey<TabsProvide> = Symbol(name);

export default defineComponent({
  name,

  props: tabsProps,

  emits: [
    'click',
    'change',
    'scroll',
    'disabled',
    'rendered',
    'click-tab',
    'update:active',
  ],

  setup(props, { emit, slots }) {
    if (process.env.NODE_ENV !== 'production') {
      const props = getCurrentInstance()?.vnode?.props;
      if (props && 'onClick' in props) {
        console.warn(
          '[Vant] Tabs: "click" event is deprecated, using "click-tab" instead.'
        );
      }
      if (props && 'onDisabled' in props) {
        console.warn(
          '[Vant] Tabs: "disabled" event is deprecated, using "click-tab" instead.'
        );
      }
    }

    let tabHeight: number;
    let lockScroll: boolean;
    let stickyFixed: boolean;

    const root = ref<HTMLElement>();
    const navRef = ref<HTMLElement>();
    const wrapRef = ref<HTMLElement>();
    const contentRef = ref<ComponentInstance>();

    const id = useId();
    const scroller = useScrollParent(root);
    const [titleRefs, setTitleRefs] = useRefs<ComponentInstance>();
    const { children, linkChildren } = useChildren(TABS_KEY);

    const state = reactive({
      inited: false,
      position: '',
      lineStyle: {} as CSSProperties,
      currentIndex: -1,
    });

    // whether the nav is scrollable
    const scrollable = computed(
      () =>
        children.length > props.swipeThreshold ||
        !props.ellipsis ||
        props.shrink
    );

    const navStyle = computed(() => ({
      borderColor: props.color,
      background: props.background,
    }));

    const getTabName = (tab: ComponentInstance, index: number): Numeric =>
      tab.name ?? index;

    const currentName = computed(() => {
      const activeTab = children[state.currentIndex];

      if (activeTab) {
        return getTabName(activeTab, state.currentIndex);
      }
    });

    const offsetTopPx = computed(() => unitToPx(props.offsetTop));

    const scrollOffset = computed(() => {
      if (props.sticky) {
        return offsetTopPx.value + tabHeight;
      }
      return 0;
    });

    // scroll active tab into view
    const scrollIntoView = (immediate?: boolean) => {
      const nav = navRef.value;
      const titles = titleRefs.value;

      if (!scrollable.value || !nav || !titles || !titles[state.currentIndex]) {
        return;
      }

      const title = titles[state.currentIndex].$el;
      const to = title.offsetLeft - (nav.offsetWidth - title.offsetWidth) / 2;

      scrollLeftTo(nav, to, immediate ? 0 : +props.duration);
    };

    // update nav bar style
    const setLine = () => {
      const shouldAnimate = state.inited;

      nextTick(() => {
        const titles = titleRefs.value;

        if (
          !titles ||
          !titles[state.currentIndex] ||
          props.type !== 'line' ||
          isHidden(root.value!)
        ) {
          return;
        }

        const title = titles[state.currentIndex].$el;
        const { lineWidth, lineHeight } = props;
        const left = title.offsetLeft + title.offsetWidth / 2;

        const lineStyle: CSSProperties = {
          width: addUnit(lineWidth),
          backgroundColor: props.color,
          transform: `translateX(${left}px) translateX(-50%)`,
        };

        if (shouldAnimate) {
          lineStyle.transitionDuration = `${props.duration}s`;
        }

        if (isDef(lineHeight)) {
          const height = addUnit(lineHeight);
          lineStyle.height = height;
          lineStyle.borderRadius = height;
        }

        state.lineStyle = lineStyle;
      });
    };

    const findAvailableTab = (index: number) => {
      const diff = index < state.currentIndex ? -1 : 1;

      while (index >= 0 && index < children.length) {
        if (!children[index].disabled) {
          return index;
        }

        index += diff;
      }
    };

    const setCurrentIndex = (
      currentIndex: number,
      skipScrollIntoView?: boolean
    ) => {
      const newIndex = findAvailableTab(currentIndex);

      if (!isDef(newIndex)) {
        return;
      }

      const newTab = children[newIndex];
      const newName = getTabName(newTab, newIndex);
      const shouldEmitChange = state.currentIndex !== null;

      if (state.currentIndex !== newIndex) {
        state.currentIndex = newIndex;

        if (!skipScrollIntoView) {
          scrollIntoView();
        }
        setLine();
      }

      if (newName !== props.active) {
        emit('update:active', newName);

        if (shouldEmitChange) {
          emit('change', newName, newTab.title);
        }
      }

      // scroll to correct position
      if (stickyFixed && !props.scrollspy) {
        setRootScrollTop(
          Math.ceil(getElementTop(root.value!) - offsetTopPx.value)
        );
      }
    };

    // correct the index of active tab
    const setCurrentIndexByName = (
      name: Numeric,
      skipScrollIntoView?: boolean
    ) => {
      const matched = children.find(
        (tab, index) => getTabName(tab, index) === name
      );

      const index = matched ? children.indexOf(matched) : 0;
      setCurrentIndex(index, skipScrollIntoView);
    };

    const scrollToCurrentContent = (immediate = false) => {
      if (props.scrollspy) {
        const target = children[state.currentIndex].$el;

        if (target && scroller.value) {
          const to = getElementTop(target, scroller.value) - scrollOffset.value;

          lockScroll = true;
          scrollTopTo(
            scroller.value,
            to,
            immediate ? 0 : +props.duration,
            () => {
              lockScroll = false;
            }
          );
        }
      }
    };

    // emit event when clicked
    const onClickTab = (
      item: ComponentInstance,
      index: number,
      event: MouseEvent
    ) => {
      const { title, disabled } = children[index];
      const name = getTabName(children[index], index);

      if (disabled) {
        // @deprecated
        // should be removed in next major version
        emit('disabled', name, title);
      } else {
        callInterceptor(props.beforeChange, {
          args: [name],
          done: () => {
            setCurrentIndex(index);
            scrollToCurrentContent();
          },
        });

        // @deprecated
        // should be removed in next major version
        emit('click', name, title);

        route(item as ComponentPublicInstance<RouteProps>);
      }

      emit('click-tab', {
        name,
        title,
        event,
        disabled,
      });
    };

    const onStickyScroll = (params: {
      isFixed: boolean;
      scrollTop: number;
    }) => {
      stickyFixed = params.isFixed;
      emit('scroll', params);
    };

    const scrollTo = (name: Numeric) => {
      nextTick(() => {
        setCurrentIndexByName(name);
        scrollToCurrentContent(true);
      });
    };

    const getCurrentIndexOnScroll = () => {
      for (let index = 0; index < children.length; index++) {
        const { top } = useRect(children[index].$el);

        if (top > scrollOffset.value) {
          return index === 0 ? 0 : index - 1;
        }
      }

      return children.length - 1;
    };

    const onScroll = () => {
      if (props.scrollspy && !lockScroll) {
        const index = getCurrentIndexOnScroll();
        setCurrentIndex(index);
      }
    };

    const renderNav = () =>
      children.map((item, index) => (
        <TabsTitle
          key={item.id}
          v-slots={{ title: item.$slots.title }}
          id={`${id}-${index}`}
          ref={setTitleRefs(index)}
          type={props.type}
          color={props.color}
          style={item.titleStyle}
          class={item.titleClass}
          shrink={props.shrink}
          isActive={index === state.currentIndex}
          controls={item.id}
          scrollable={scrollable.value}
          activeColor={props.titleActiveColor}
          inactiveColor={props.titleInactiveColor}
          onClick={(event: MouseEvent) => onClickTab(item, index, event)}
          {...pick(item, [
            'dot',
            'badge',
            'title',
            'disabled',
            'showZeroBadge',
          ])}
        />
      ));

    const renderLine = () => {
      if (props.type === 'line' && children.length) {
        return <div class={bem('line')} style={state.lineStyle} />;
      }
    };

    const renderHeader = () => {
      const { type, border, sticky } = props;

      const Header = [
        <div
          ref={sticky ? undefined : wrapRef}
          class={[
            bem('wrap'),
            { [BORDER_TOP_BOTTOM]: type === 'line' && border },
          ]}
        >
          <div
            ref={navRef}
            role="tablist"
            class={bem('nav', [
              type,
              { shrink: props.shrink, complete: scrollable.value },
            ])}
            style={navStyle.value}
            aria-orientation="horizontal"
          >
            {slots['nav-left']?.()}
            {renderNav()}
            {renderLine()}
            {slots['nav-right']?.()}
          </div>
        </div>,
        slots['nav-bottom']?.(),
      ];

      if (sticky) {
        return <div ref={wrapRef}>{Header}</div>;
      }
      return Header;
    };

    watch([() => props.color, windowWidth], setLine);

    watch(
      () => props.active,
      (value) => {
        if (value !== currentName.value) {
          setCurrentIndexByName(value);
        }
      }
    );

    watch(
      () => children.length,
      () => {
        if (state.inited) {
          setCurrentIndexByName(props.active);
          setLine();
          nextTick(() => {
            scrollIntoView(true);
          });
        }
      }
    );

    const init = () => {
      setCurrentIndexByName(props.active, true);
      nextTick(() => {
        state.inited = true;
        if (wrapRef.value) {
          tabHeight = useRect(wrapRef.value).height;
        }
        scrollIntoView(true);
      });
    };

    const onRendered = (name: Numeric, title?: string) =>
      emit('rendered', name, title);

    const resize = () => {
      setLine();
      nextTick(() => contentRef.value?.swipeRef.value?.resize());
    };

    useExpose({
      resize,
      scrollTo,
    });

    onActivated(setLine);
    onPopupReopen(setLine);
    onMountedOrActivated(init);
    useEventListener('scroll', onScroll, {
      target: scroller,
      passive: true,
    });

    linkChildren({
      id,
      props,
      setLine,
      onRendered,
      currentName,
      scrollIntoView,
    });

    return () => (
      <div ref={root} class={bem([props.type])}>
        {props.sticky ? (
          <Sticky
            container={root.value}
            offsetTop={offsetTopPx.value}
            onScroll={onStickyScroll}
          >
            {renderHeader()}
          </Sticky>
        ) : (
          renderHeader()
        )}
        <TabsContent
          ref={contentRef}
          count={children.length}
          inited={state.inited}
          animated={props.animated}
          duration={props.duration}
          swipeable={props.swipeable}
          lazyRender={props.lazyRender}
          currentIndex={state.currentIndex}
          onChange={setCurrentIndex}
        >
          {slots.default?.()}
        </TabsContent>
      </div>
    );
  },
});
