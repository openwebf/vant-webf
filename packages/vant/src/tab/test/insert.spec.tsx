import { defineComponent } from 'vue';
import { mount, later } from '../../../test';
import { Tab } from '..';
import { Tabs } from '../../tabs';

test('should render correctly after inserting a tab', async () => {
  const wrapper = mount({
    render() {
      return (
        <Tabs v-model:active={this.active}>
          <Tab title="1">1</Tab>
          {this.insert && (
            <div>
              <Tab title="2">2</Tab>
            </div>
          )}
          <Tab title="3">3</Tab>
        </Tabs>
      );
    },
    data() {
      return {
        insert: false,
        active: 1,
      };
    },
  });

  await later();
  await wrapper.setData({ insert: true });
  await later();
  expect(wrapper.html()).toMatchSnapshot();
});

test('should render correctly after inserting a tab with name', async () => {
  const onChange = jest.fn();
  const wrapper = mount({
    render() {
      return (
        <Tabs v-model:active={this.active} onChange={onChange}>
          {this.insert && (
            <Tab title="bar" name="bar">
              bar
            </Tab>
          )}
          <Tab title="foo" name="foo">
            foo
          </Tab>
        </Tabs>
      );
    },
    data() {
      return {
        insert: false,
        active: 'foo',
      };
    },
  });

  await later();
  await wrapper.setData({ insert: true });
  await later();
  expect(wrapper.html()).toMatchSnapshot();
  expect(onChange).toHaveBeenCalledTimes(0);
});

test('should render Tab inside a component correctly', async () => {
  const MyTab = defineComponent({
    render() {
      return <Tab title="2">2</Tab>;
    },
  });

  const wrapper = mount({
    render() {
      return (
        <Tabs active={1}>
          <Tab title="1">1</Tab>
          <MyTab />
          <Tab title="3">3</Tab>
        </Tabs>
      );
    },
  });

  await later();
  expect(wrapper.html()).toMatchSnapshot();
});

test('should render nav-bottom slot correctly', async () => {
  const wrapper = mount(Tabs, {
    slots: {
      'nav-bottom': () => 'Nav Bottom',
    },
  });

  await later();
  expect(wrapper.html()).toMatchSnapshot();
});
