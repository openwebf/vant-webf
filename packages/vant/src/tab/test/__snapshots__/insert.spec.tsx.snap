// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Tab inside a component correctly 1`] = `
<div class="van-tabs van-tabs--line">
  <div class="van-tabs__wrap">
    <div role="tablist"
         class="van-tabs__nav van-tabs__nav--line"
         aria-orientation="horizontal"
    >
      <div id="van-tabs-0"
           role="tab"
           class="van-tab van-tab--line"
           tabindex="-1"
           aria-selected="false"
           aria-controls="van-tab"
      >
        <span class="van-tab__text van-tab__text--ellipsis">
          1
        </span>
      </div>
      <div id="van-tabs-1"
           role="tab"
           class="van-tab van-tab--line van-tab--active"
           tabindex="0"
           aria-selected="true"
           aria-controls="van-tab"
      >
        <span class="van-tab__text van-tab__text--ellipsis">
          2
        </span>
      </div>
      <div id="van-tabs-2"
           role="tab"
           class="van-tab van-tab--line"
           tabindex="-1"
           aria-selected="false"
           aria-controls="van-tab"
      >
        <span class="van-tab__text van-tab__text--ellipsis">
          3
        </span>
      </div>
      <div class="van-tabs__line"
           style="transform: translateX(50px) translateX(-50%);"
      >
      </div>
    </div>
  </div>
  <div class="van-tabs__content">
    <div id="van-tab"
         role="tabpanel"
         class="van-tab__panel"
         tabindex="-1"
         aria-labelledby="van-tabs-0"
         style="display: none;"
    >
    </div>
    <div id="van-tab"
         role="tabpanel"
         class="van-tab__panel"
         tabindex="0"
         aria-labelledby="van-tabs-1"
         style
    >
      2
    </div>
    <div id="van-tab"
         role="tabpanel"
         class="van-tab__panel"
         tabindex="-1"
         aria-labelledby="van-tabs-2"
         style="display: none;"
    >
    </div>
  </div>
</div>
`;

exports[`should render correctly after inserting a tab 1`] = `
<div class="van-tabs van-tabs--line">
  <div class="van-tabs__wrap">
    <div role="tablist"
         class="van-tabs__nav van-tabs__nav--line"
         aria-orientation="horizontal"
    >
      <div id="van-tabs-0"
           role="tab"
           class="van-tab van-tab--line"
           tabindex="-1"
           aria-selected="false"
           aria-controls="van-tab"
      >
        <span class="van-tab__text van-tab__text--ellipsis">
          1
        </span>
      </div>
      <div id="van-tabs-1"
           role="tab"
           class="van-tab van-tab--line van-tab--active"
           tabindex="0"
           aria-selected="true"
           aria-controls="van-tab"
      >
        <span class="van-tab__text van-tab__text--ellipsis">
          2
        </span>
      </div>
      <div id="van-tabs-2"
           role="tab"
           class="van-tab van-tab--line"
           tabindex="-1"
           aria-selected="false"
           aria-controls="van-tab"
      >
        <span class="van-tab__text van-tab__text--ellipsis">
          3
        </span>
      </div>
      <div class="van-tabs__line"
           style="transform: translateX(50px) translateX(-50%); transition-duration: 0.3s;"
      >
      </div>
    </div>
  </div>
  <div class="van-tabs__content">
    <div id="van-tab"
         role="tabpanel"
         class="van-tab__panel"
         tabindex="-1"
         aria-labelledby="van-tabs-0"
         style="display: none;"
    >
    </div>
    <div>
      <div id="van-tab"
           role="tabpanel"
           class="van-tab__panel"
           tabindex="0"
           aria-labelledby="van-tabs-1"
      >
        2
      </div>
    </div>
    <div id="van-tab"
         role="tabpanel"
         class="van-tab__panel"
         tabindex="-1"
         aria-labelledby="van-tabs-2"
         style="display: none;"
    >
      3
    </div>
  </div>
</div>
`;

exports[`should render correctly after inserting a tab with name 1`] = `
<div class="van-tabs van-tabs--line">
  <div class="van-tabs__wrap">
    <div role="tablist"
         class="van-tabs__nav van-tabs__nav--line"
         aria-orientation="horizontal"
    >
      <div id="van-tabs-0"
           role="tab"
           class="van-tab van-tab--line"
           tabindex="-1"
           aria-selected="false"
           aria-controls="van-tab"
      >
        <span class="van-tab__text van-tab__text--ellipsis">
          bar
        </span>
      </div>
      <div id="van-tabs-1"
           role="tab"
           class="van-tab van-tab--line van-tab--active"
           tabindex="0"
           aria-selected="true"
           aria-controls="van-tab"
      >
        <span class="van-tab__text van-tab__text--ellipsis">
          foo
        </span>
      </div>
      <div class="van-tabs__line"
           style="transform: translateX(50px) translateX(-50%); transition-duration: 0.3s;"
      >
      </div>
    </div>
  </div>
  <div class="van-tabs__content">
    <div id="van-tab"
         role="tabpanel"
         class="van-tab__panel"
         tabindex="-1"
         aria-labelledby="van-tabs-0"
         style="display: none;"
    >
      bar
    </div>
    <div id="van-tab"
         role="tabpanel"
         class="van-tab__panel"
         tabindex="0"
         aria-labelledby="van-tabs-1"
         style
    >
      foo
    </div>
  </div>
</div>
`;

exports[`should render nav-bottom slot correctly 1`] = `
<div class="van-tabs van-tabs--line">
  <div class="van-tabs__wrap">
    <div role="tablist"
         class="van-tabs__nav van-tabs__nav--line"
         aria-orientation="horizontal"
    >
    </div>
  </div>
  Nav Bottom
  <div class="van-tabs__content">
  </div>
</div>
`;
